x/wasm/internal/keeper/keeper.go

```diff
@@ -212,6 +212,55 @@ func (k Keeper) Execute(ctx sdk.Context, contractAddress sdk.AccAddress, caller
 	return value, nil
 }
 
+// Migrate allows to upgrade a contract to a new code with data migration.
+func (k Keeper) Migrate(ctx sdk.Context, contractAddress sdk.AccAddress, caller sdk.AccAddress, newCodeID uint64, msg []byte) (*sdk.Result, error) {
+	contractInfo := k.GetContractInfo(ctx, contractAddress)
+	if contractInfo == nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, "unknown contract")
+	}
+	if contractInfo.Admin == nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "migration not supported by this contract")
+	}
+	if !contractInfo.Admin.Equals(caller) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrUnauthorized, "no permission")
+	}
+	newCodeInfo := k.GetCodeInfo(ctx, newCodeID)
+	if newCodeInfo == nil {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "unknown code")
+	}
+
+	var deposit sdk.Coins // todo: does it make sense to pass nil value here?
+	params := types.NewEnv(ctx, caller, deposit, contractAddress)
+
+	// prepare querier
+	querier := QueryHandler{
+		Ctx:     ctx,
+		Plugins: k.queryPlugins,
+	}
+
+	prefixStoreKey := types.GetContractStorePrefixKey(contractAddress)
+	prefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), prefixStoreKey)
+	gas := gasForContract(ctx)
+	res, execErr := k.wasmer.Migrate(newCodeInfo.CodeHash, params, msg, prefixStore, cosmwasmAPI, querier, gas)
+	if execErr != nil {
+		// TODO: wasmer doesn't return wasm gas used on error. we should consume it (for error on metering failure)
+		// Note: OutOfGas panics (from storage) are caught by go-cosmwasm, subtract one more gas to check if
+		// this contract died due to gas limit in Storage
+		consumeGas(ctx, GasMultiplier)
+		return nil, sdkerrors.Wrap(types.ErrExecuteFailed, execErr.Error())
+	}
+	consumeGas(ctx, res.GasUsed)
+
+	// emit all events from this contract migration itself
+	value := types.CosmosResult(*res, contractAddress)
+	ctx.EventManager().EmitEvents(value.Events)
+	value.Events = nil
+
+	contractInfo.CodeID = newCodeID
+	k.setContractInfo(ctx, contractAddress, *contractInfo)
```

Comment:
After doing the update (migration finished), we should optionally handle dispatching returned messages (like in Execute): ` k.dispatchMessages(ctx, contractAddress, res.Messages)`

This let's us trigger external actions upon migrate. Eg. the contract could delete state and send all remaining bank balance to an address set in the MigrateMsg as  a way to self-destruct.